### ISSUE TITLE
Help with kerastuneR

### ISSUE THREAD

**Issue by ssvargas** [EagerAI/kerastuneR#48]
I have a basic NN architecture already coded, but I want to tune some hyperparameters on it by using kerastuneR library, and I am not quite sure on how to code it.

The architecture is as follow:
```
# Define the model architecture
model <- keras_model_sequential() %>%
  layer_conv_2d(name = input1[1],
                          filters = TO BE TUNED,
                          kernel_size = c(3, 3),
                          activation = "relu",
                          kernel_regularizer = regularizer_l2(TO BE TUNED),
                          input_shape = c(shape1, shape2, 1)) %>%
  layer_max_pooling_2d(pool_size = c(2, 2)) %>%

  layer_flatten() %>%

  layer_dense(name = input1[2], units = TO BE TUNED,
                      activation = "relu") %>%
  layer_dropout(rate = TO BE TUNED) %>%

  layer_dense(name = input1[3], units = TO BE TUNED,
                      activation = "relu") %>%
  layer_dropout(rate = TO BE TUNED) %>%

  layer_dense(name = input1[4], units = TO BE TUNED,
                      activation = "relu") %>%
  layer_dropout(rate = TO BE TUNED) %>%

  layer_dense(name = input1[5], units = 2, activation = "softmax")

# Create an optimizer
optimizer <- optimizer_rmsprop(learning_rate = TO BE TUNED,
                                                     rho = TO BE TUNED,
                                                     epsilon = TO BE TUNED
                                                     )

# Compile the model
model %>% compile(loss = "binary_crossentropy",
                                  optimizer = optimizer,
                                  metrics = "accuracy"
                                  )

# Fit the model
history <- model %>% fit(x = xtrain,
                                          y = ytrain,
                                          validation_data = list(xval, yval),
                                          epochs = TO BE TUNED,
                                          batch_size = TO BE TUNED,
                                          verbose = 1
                                          )
```
The aim of the tuning process is to find the best combination of hyperparameters TO BE TUNED, in orther to achieve the best, consistent accuracy in both, the training and the validation data, and of course to save that model.

**Comment by turgut090** [comment]
Hi, can you put your code into function? Please see below example:

```
library(keras)
library(tensorflow)
library(kerastuneR)

build_model = function(hp) {
  
  model = keras_model_sequential()
  model %>% layer_dense(units = hp$Int('units',
                                     min_value = 32,
                                     max_value = 512,
                                     step=  32),input_shape = ncol(x_data),
                        activation =  'relu') %>%
    layer_dense(units = 1, activation = 'softmax') %>%
    compile(
      optimizer = tf$keras$optimizers$Adam(
        hp$Choice('learning_rate',
                  values=c(1e-2, 1e-3, 1e-4))),
      loss = 'binary_crossentropy',
      metrics = 'accuracy')
  return(model)
}

tuner = RandomSearch(
    build_model,
    objective = 'val_accuracy',
    max_trials = 5,
    executions_per_trial = 3,
    directory = 'my_dir',
    project_name = 'helloworld')

tuner %>% fit_tuner(x_data,y_data,
                    epochs = 5, 
                    validation_data = list(x_data2,y_data2))
```

**Comment by ssvargas** [comment]
Dear Turgut,

Thank you so much for your answer. In fact, I have put the model into the function, and the code runs. However, instead of improving the algorithm performance, after 10 trials, when I run get_best_models() function, the "best" algorithm has an accuracy of 0.5, less than the first trial performance.

This is how I wrote the script. If you have any suggestion or thought about the error, I will appreciate so much your comments.

# Define the hyperparamenter ranges
hp <- HyperParameters()

hp$Choice("filters2d", 2^c(4:8))  # the filters in the 2d layer
hp$Choice("regularizer", 10^c(-3:-2))  # the regularizer
hp$Choice("units1", 2^c(3:7))  # the number of units in layer 1
hp$Choice("dropout1", 0.05 + c(0:3)*0.05)  # the dropout rate in layer 1
hp$Choice("units2", 2^c(3:7))  # the number of units in layer 2
hp$Choice("dropout2", 0.05 + c(0:3)*0.05)  # the dropout rate in layer 2
hp$Choice("units3", 2^c(3:7))  # the number of units in layer 3
hp$Choice("dropout3", 0.05 + c(0:3)*0.05)  # the dropout rate in layer 3
hp$Choice("units4", 2^c(3:7))  # the number of units in layer 4
hp$Choice("dropout4", 0.05 + c(0:3)*0.05)  # the dropout rate in layer 4
hp$Choice("lr", 0.003 + c(0:4)*0.006)  # the learning rate
hp$Choice("rho", 0.85 + c(0:2)*0.05)  # the rho
hp$Choice("eps", 10^c(-9:-8))  # the epsilon

# Define the model training strategy
nnarchitecture <- function (hp) {
  
  # The architecture
  model <- keras_model_sequential() %>%
    
    layer_conv_2d(name = "layer2d",
                  filters = hp$get("filters2d"),
                  kernel_size = c(3, 3),
                  activation = "relu",
                  kernel_regularizer = regularizer_l2(hp$get("regularizer")),
                  input_shape = c(shape1, shape2, 1)) %>%
    layer_max_pooling_2d(pool_size = c(2, 2)) %>%
    
    layer_flatten() %>%
    
    ###
    layer_dense(name = "layer1",
                units = hp$get("units1"),
                activation = "relu") %>%
    layer_dropout(rate = hp$get("dropout1")) %>%
    
    layer_dense(name = "layer2",
                units = hp$get("units2"),
                activation = "relu") %>%
    layer_dropout(rate = hp$get("dropout2")) %>%
    
    layer_dense(name = "layer3",
                units = hp$get("units3"),
                activation = "relu") %>%
    layer_dropout(rate = hp$get("dropout3")) %>%
    
    layer_dense(name = "layer4",
                units = hp$get("units4"),
                activation = "relu") %>%
    layer_dropout(rate = hp$get("dropout4")) %>%
    ###
    
    layer_dense(name = "layerout", units = 2, activation = "softmax")
  
  # The optimizer
  optimizer <- optimizer_rmsprop(learning_rate = hp$get("lr"),
                                 rho = hp$get("rho"),
                                 epsilon = hp$get("eps"))
  
  # Compile the model
  model %>% compile(loss = "binary_crossentropy",
                    optimizer = optimizer,
                    metrics = "accuracy")
  
  return(model)
}

# Define the tune strategy
tuner <- RandomSearch(hypermodel = nnarchitecture,
                      max_trials = 10,
                      hyperparameters = hp,
                      tune_new_entries = TRUE,
                      objective = "val_accuracy",
                      directory = "./Outputs/",
                      project_name = "model")

# Train the model
tuner %>% fit_tuner(x = xtrain,
                    y = ytrain,
                    epochs = 5,
                    validation_data = list(xval, yval),
                    batch_size = 16)

Best regards

**Comment by turgut090** [comment]
Can you share your data as well? Is it artificial data or real-world data? Try to implement tuning for example for iris dataset.
